File: ejemplos/general/01-lista-de-tareas-cli/ejemplo_menu_interactivo.py
"""
Ejemplo: Mini gestor de contactos con menú interactivo y persistencia JSON

Qué enseña este ejemplo:
  - Cómo crear un loop de menú principal con while True + break
  - Cómo usar input() para leer texto y números del usuario
  - Cómo guardar y cargar datos con json.dump / json.load
  - Cómo actualizar un elemento específico de una lista (modificar un diccionario)
  - Cómo eliminar un elemento usando list.pop() o del
  - Cómo manejar el caso en que el archivo aún no existe

Para tu reto (lista de tareas):
  Aplica el mismo patrón de loop + input() + guardado JSON, pero en vez de 
  contactos (nombre/teléfono) trabajarás con tareas (texto/completado).
  - Tu "agregar" será para nuevas tareas.
  - Tu "actualizar" será tu "marcar como completada".
  - Tu "eliminar" funcionará igual que eliminar un contacto.

Glosario de términos "raros":
  with open(...) as f — Abre un archivo de forma segura. "f" es solo una variable 
                        corta para "file" (archivo). Si hay error, cierra el archivo solo.
  "r" o "w"           — Modos al abrir un archivo: "r" (read = leer), "w" (write = escribir).
  encoding="utf-8"    — Asegura que los acentos y caracteres especiales (ñ, á) se guarden bien.
  json.load()         — Convierte texto JSON de un archivo a una lista/diccionario en Python.
  json.dump()         — Convierte una lista/diccionario en Python a texto JSON en un archivo.
"""

import json
import os

ARCHIVO = "contactos.json"


def cargar_contactos():
    """
    Carga la lista de contactos desde el archivo JSON.
    Si el archivo no existe (primera vez), devuelve una lista vacía.
    """
    if not os.path.exists(ARCHIVO):
        return []

    with open(ARCHIVO, "r", encoding="utf-8") as f:
        return json.load(f)


def guardar_contactos(contactos):
    """Guarda la lista de contactos en el archivo JSON."""
    with open(ARCHIVO, "w", encoding="utf-8") as f:
        json.dump(contactos, f, indent=4, ensure_ascii=False)


def mostrar_contactos(contactos):
    if not contactos:
        print("No hay contactos guardados todavía.")
        return
    print("\n--- Tus contactos ---")
    for i, contacto in enumerate(contactos, start=1):
        print(f"  {i}. {contacto['nombre']} — {contacto['telefono']}")


def agregar_contacto(contactos):
    nombre = input("Nombre: ").strip()
    telefono = input("Teléfono: ").strip()
    if nombre:
        contactos.append({"nombre": nombre, "telefono": telefono})
        guardar_contactos(contactos)
        print(f"✓ Contacto '{nombre}' guardado.")
    else:
        print("El nombre no puede estar vacío.")


def actualizar_contacto(contactos):
    """Demuestra cómo modificar un elemento existente en la lista."""
    mostrar_contactos(contactos)
    if not contactos:
        return

    try:
        idx = int(input("\nNúmero del contacto a actualizar: ")) - 1
        if 0 <= idx < len(contactos):
            nuevo_tel = input("Nuevo teléfono: ").strip()
            # Modificamos el valor dentro del diccionario
            contactos[idx]["telefono"] = nuevo_tel
            guardar_contactos(contactos)
            print("✓ Contacto actualizado.")
        else:
            print("Número inválido.")
    except ValueError:
        print("Por favor, ingresa un número válido.")


def eliminar_contacto(contactos):
    """Demuestra cómo eliminar un elemento de la lista usando pop()."""
    mostrar_contactos(contactos)
    if not contactos:
        return

    try:
        idx = int(input("\nNúmero del contacto a eliminar: ")) - 1
        if 0 <= idx < len(contactos):
            # pop() remueve y devuelve el elemento en ese índice
            eliminado = contactos.pop(idx)
            guardar_contactos(contactos)
            print(f"✓ Contacto '{eliminado['nombre']}' eliminado.")
        else:
            print("Número inválido.")
    except ValueError:
        print("Por favor, ingresa un número válido.")


def main():
    contactos = cargar_contactos()

    while True:
        print("\n=== Gestor de Contactos ===")
        print("1. Ver contactos")
        print("2. Agregar contacto")
        print("3. Actualizar teléfono")
        print("4. Eliminar contacto")
        print("5. Salir")

        opcion = input("\nElige una opción (1-5): ").strip()

        if opcion == "1":
            mostrar_contactos(contactos)
        elif opcion == "2":
            agregar_contacto(contactos)
        elif opcion == "3":
            actualizar_contacto(contactos)
        elif opcion == "4":
            eliminar_contacto(contactos)
        elif opcion == "5":
            print("¡Hasta luego!")
            break
        else:
            print("Opción no reconocida, intenta de nuevo.")


if __name__ == "__main__":
    main()

File: archivo-historico/ejemplos/general/01-lista-de-tareas-cli/ejemplo_menu_interactivo.go
// Ejemplo: Mini gestor de contactos con menú interactivo y persistencia JSON en Go
//
// Qué enseña este ejemplo:
//   - Cómo crear un loop con for {}
//   - Cómo usar bufio.NewScanner para leer líneas de la terminal
//   - Cómo leer y guardar archivos con os.ReadFile y os.WriteFile
//   - Cómo usar encoding/json para convertir structs de Go a JSON y viceversa
//
// Glosario de términos "raros":
//   bufio.NewScanner — Crea un "escáner" que facilita leer texto de la consola línea por línea.
//   os.ReadFile      — Lee todo el contenido de un archivo de una sola vez.
//   os.WriteFile     — Guarda datos en un archivo, sobreescribiéndolo.
//   json.Marshal     — (De "marshall") Convierte tus variables de Go a texto JSON.
//   json.Unmarshal   — Convierte texto JSON de vuelta a variables de Go.
//   0644             — Permisos del archivo (lectura/escritura para ti, lectura para otros).

package main

import (
	"bufio"
	"encoding/json"
	"fmt"
	"os"
	"strconv"
	"strings"
)

const ARCHIVO = "contactos.json"

// Contacto representa un elemento de nuestra lista.
// Las etiquetas `json:"nombre"` indican cómo se llamará el campo en el archivo.
type Contacto struct {
	Nombre   string `json:"nombre"`
	Telefono string `json:"telefono"`
}

func cargarContactos() []Contacto {
	// Intentamos leer el archivo
	data, err := os.ReadFile(ARCHIVO)
	if err != nil {
		// Si hay un error (ej. el archivo no existe), devolvemos lista vacía
		return []Contacto{}
	}

	var contactos []Contacto
	// Convertimos el JSON a la lista de structs
	json.Unmarshal(data, &contactos)
	return contactos
}

func guardarContactos(contactos []Contacto) {
	// MarshalIndent hace que el JSON sea legible (con espacios)
	data, _ := json.MarshalIndent(contactos, "", "    ")
	os.WriteFile(ARCHIVO, data, 0644)
}

func mostrarContactos(contactos []Contacto) {
	if len(contactos) == 0 {
		fmt.Println("No hay contactos guardados todavía.")
		return
	}
	fmt.Println("\n--- Tus contactos ---")
	for i, contacto := range contactos {
		fmt.Printf("  %d. %s — %s\n", i+1, contacto.Nombre, contacto.Telefono)
	}
}

// Función auxiliar para leer texto
func preguntar(prompt string, scanner *bufio.Scanner) string {
	fmt.Print(prompt)
	scanner.Scan()
	return strings.TrimSpace(scanner.Text())
}

func main() {
	contactos := cargarContactos()
	scanner := bufio.NewScanner(os.Stdin)

	for {
		fmt.Println("\n=== Gestor de Contactos ===")
		fmt.Println("1. Ver contactos")
		fmt.Println("2. Agregar contacto")
		fmt.Println("3. Actualizar teléfono")
		fmt.Println("4. Eliminar contacto")
		fmt.Println("5. Salir")

		opcion := preguntar("\nElige una opción (1-5): ", scanner)

		switch opcion {
		case "1":
			mostrarContactos(contactos)

		case "2":
			nombre := preguntar("Nombre: ", scanner)
			telefono := preguntar("Teléfono: ", scanner)

			if nombre != "" {
				contactos = append(contactos, Contacto{Nombre: nombre, Telefono: telefono})
				guardarContactos(contactos)
				fmt.Printf("✓ Contacto '%s' guardado.\n", nombre)
			} else {
				fmt.Println("El nombre no puede estar vacío.")
			}

		case "3":
			mostrarContactos(contactos)
			if len(contactos) == 0 {
				continue
			}

			idxStr := preguntar("\nNúmero del contacto a actualizar: ", scanner)
			idx, err := strconv.Atoi(idxStr)
			idx -= 1 // Ajustamos para índice 0

			if err == nil && idx >= 0 && idx < len(contactos) {
				nuevoTel := preguntar("Nuevo teléfono: ", scanner)
				contactos[idx].Telefono = nuevoTel
				guardarContactos(contactos)
				fmt.Println("✓ Contacto actualizado.")
			} else {
				fmt.Println("Número inválido.")
			}

		case "4":
			mostrarContactos(contactos)
			if len(contactos) == 0 {
				continue
			}

			idxStr := preguntar("\nNúmero del contacto a eliminar: ", scanner)
			idx, err := strconv.Atoi(idxStr)
			idx -= 1

			if err == nil && idx >= 0 && idx < len(contactos) {
				eliminado := contactos[idx]
				// Magia para eliminar un elemento de un slice en Go:
				contactos = append(contactos[:idx], contactos[idx+1:]...)
				guardarContactos(contactos)
				fmt.Printf("✓ Contacto '%s' eliminado.\n", eliminado.Nombre)
			} else {
				fmt.Println("Número inválido.")
			}

		case "5":
			fmt.Println("¡Hasta luego!")
			return

		default:
			fmt.Println("Opción no reconocida, intenta de nuevo.")
		}
	}
}
